content/post/74-putting-npm-on-ipfs-part-2.md

```diff
@@ -0,0 +1,90 @@
+Swarm listening on /ip4/127.0.0.1/tcp/49905/ipfs/QmZ7vEvXRdTVipb9o2p2Cmt3s1S8rqbo2ohscTjvTLgnpP
+🗂️  Loading registry index from https://registry.js.ipfs.io
+☎️  Dialling registry mirror /ip4/35.178.192.119/tcp/10040/ipfs/QmfKqxieE71QoAchNk5e2MKmvWKjGdUnSifHqq1xZLEzyn,/dns4/registry.js.ipfs.io/tcp/10040/ipfs/QmfKqxieE71QoAchNk5e2MKmvWKjGdUnSifHqq1xZLEzyn
+📱️ Connected to registry
+👩‍🚀 Starting local proxy
+🚀 Server running on port 49935
+🎁 Installing dependencies with /Users/alex/.nvm/versions/node/v10.15.1/bin/yarn
+yarn install v1.13.0
+...
+```
+
+## 🙋 How does it work?
+
+Behind the scenes `ipfs-npm` first starts a local http server and configures npm/Yarn to use it as the registry.  When a module's [packument](https://github.com/zkat/pacote/tree/33c53cf10b080e78182bccc56ec1d5126f8b627e#packument) is requested, it uses the same package metadata as [registry.js.ipfs.io](https://registry.js.ipfs.io) to satisfy the request, also checking with the central registry for any updated versions and including them if available.
+
+`npm`/`yarn` uses the package metadata to select a tarball to request that satisfies the [semver](https://semver.org/) range in the developer's `package.json` for the dependency. Once the tarball request is recieved, it uses the requested URL to look up the relevant CID in the module's metadata - if a CID is present, it uses IPFS to request the content, otherwise it downloads it from the public npm registry, adds it to IPFS and stores the CID for the next time it's requested.
+
+You can use `ipfs-npm` with an existing IPFS node running on your computer or remotely, or fall back to the default configuration which is to run an IPFS node for the duration of the install.
+
+![Request sequence](/74-putting-npm-on-ipfs-part-2/ipfs-npm-sequence.png)
+
+If this sounds familiar it may be because you've just read [part 1](/post/73-putting-npm-on-ipfs-part-1) - indeed it's the exact same code fulfilling the exact same function, just locally on your computer instead of on a server somewhere.
+
+The http server is necessary because at the time of writing neither npm or Yarn support IPFS as a transport - hopefully one day this won't be necessary but as it stands no http traffic leaves your machine, unless you request modules that don't have CIDs in their metadata.
+
+It also dials the ipfs-npm-registry-mirror directly to improve the speed of content resolution - this is partly because we know that the mirror will have the content we are after but also because there is no [DHT](https://ipfs.io/ipfs/QmXoypizjW3WknFiJnKLwHCnL72vedxjQkDDP1mXWo6uco/wiki/Distributed_hash_table.html) in js-IPFS (yet - it's coming!) so discovery would have to occur via the gateway nodes.
+
+## 🎁 What's next?
+
+So many things!
+
+* We'd love (love!) to get IPFS into npm and Yarn as a first-class transport option
+* Once the DHT is enabled in js-ipfs v0.35.0 we'll not need to contact the npm mirror so a point of failure will be removed and startup will be faster
```

Comment:
```suggestion
* Once the DHT is enabled in js-ipfs v0.35.0 we'll not need to contact the npm mirror so that single point of failure will be removed and startup will be faster
```